rllib/evaluation/collectors/agent_collector.py

```diff
@@ -552,6 +551,18 @@ def _unflatten_as_buffer_struct(
         """Unflattens the given to match the buffer struct format for that key."""
         if key not in self.buffer_structs:
             return data[0]
+        if key == SampleBatch.ACTIONS and not self.disable_action_flattening:
+            from ray.rllib.utils.spaces.space_utils import flatten_to_single_ndarray
+
+            actions = None
+            # for
+            for shards in data:
+                if actions is None:
+                    actions = [[] for _ in range(len(shards))]
+                for action, shard in zip(actions, shards):
+                    action.append(shard)
+            actions = [flatten_to_single_ndarray(a) for a in actions]
```

Comment:
dude this code seems odd to me. I may be missing something here but the code for this should look much simpler than this. It's not obvious what is happening here. maybe this?
```
actions = [flatten_to_single_ndarray(d)  for d in data]
```

Reply:
Wait but, the data contains the shards of the whole action, when I inspected it. 

It's structure is like:

[[[action0_shard0], [action1_shard0]], 

[[action0_shard1], [action1_shard1]]]

Reply:
by shard_0 you mean the leaf_0?

Reply:
Yep leaf is the better word for that.

Reply:
is my example right?
```
data = [np.array([[1, 2], [3, 4]]), np.array([10, 20])] --> desired action = np.array([[1, 2, 10], [3, 4, 20]])
data =  [np.array([[1], [3]]), np.array([10, 20])] --> desired action = np.array([[1, 10], [3, 20]])
```

Reply:
Yep that is what this block of code does.

Reply:
we only have 2d and 1d arrays in this method? can we get 3d arrays as well?

Reply:
in either case I think we should do this:
```
actions = np.concatenate([d.view(d.shape[0], -1) for d in data], -1)
```